diffusion/datasets/image_caption_latents.py

```diff
@@ -0,0 +1,246 @@
+        self.caption_keys = caption_keys
+        self.caption_selection_probs = caption_selection_probs
+        self.text_latent_keys = text_latent_keys
+        self.text_latent_shapes = text_latent_shapes
+        self.attention_mask_keys = attention_mask_keys
+
+    def __getitem__(self, index):
+        sample = super().__getitem__(index)
+        out = {}
+
+        # Image
+        img = sample[self.image_key]
+        if not isinstance(img, Image.Image):
+            img = Image.open(BytesIO(sample[self.image_key]))
+        if img.mode != 'RGB':
+            img = img.convert('RGB')
+        out['cond_original_size'] = torch.tensor(img.size)
+
+        # Image transforms
+        if self.crop is not None:
+            img, crop_top, crop_left = self.crop(img)
+        else:
+            crop_top, crop_left = 0, 0
+        out['cond_crops_coords_top_left'] = torch.tensor([crop_top, crop_left])
+
+        if self.transform is not None:
+            img = self.transform(img)
+        out['image'] = img
+
+        # Get the new height and width
+        if isinstance(img, torch.Tensor):
+            img_h, img_w = img.shape[-2], img.shape[-1]
+        elif isinstance(img, Image.Image):
+            img_w, img_h = img.size
+        else:
+            raise ValueError('Image after transformations must either be a PIL Image or Torch Tensor')
+        out['cond_target_size'] = torch.tensor([img_w, img_h])
+
+        # Microconditioning dropout as in Stability repo
+        # https://github.com/Stability-AI/generative-models/blob/477d8b9a7730d9b2e92b326a770c0420d00308c9/sgm/modules/encoders/modules.py#L151-L160
+        if torch.rand(1) < self.microcond_drop_prob:
+            out['cond_crops_coords_top_left'] = out['cond_crops_coords_top_left'] * 0
+        if torch.rand(1) < self.microcond_drop_prob:
+            out['cond_original_size'] = out['cond_original_size'] * 0
+        if torch.rand(1) < self.microcond_drop_prob:
+            out['cond_target_size'] = out['cond_target_size'] * 0
+
+        # Randomly select a caption according to the selection probabilities
+        caption_key = np.random.choice(self.caption_keys, p=self.caption_selection_probs)
+        # Load text latents, attention masks, and clip pooled embeddings
+        for i in range(len(self.text_latent_keys)):
+            latent_key = f'{caption_key}_{self.text_latent_keys[i]}'
+            latent_shape = self.text_latent_shapes[i]
+            attention_key = f'{caption_key}_{self.attention_mask_keys[i]}'
+
+            if torch.rand(1) < self.caption_drop_prob:
+                out[self.text_latent_keys[i]] = torch.zeros(latent_shape, dtype=torch.float16)
+                out[self.attention_mask_keys[i]] = torch.zeros(latent_shape[0])
+                if 'CLIP_LATENTS' in latent_key:
+                    out['CLIP_POOLED'] = torch.zeros(latent_shape[1])
+            else:
+                text_latent = np.frombuffer(sample[latent_key], dtype=np.float16).copy()
```

Comment:
confirming that the dtype for the latents are valid for training - ie, model expects tensors in float16 as well?

Reply:
We are planning on this, but perhaps it should be configurable

Reply:
Changed it to be configurable!